File: packages/backend/eval/generate-all97-report.py
#!/usr/bin/env python3
"""Parse mlj017-all97-run-utf8.txt and produce a markdown results report."""
import re, json, os

eval_dir = os.path.dirname(os.path.abspath(__file__))
raw_path = os.path.join(eval_dir, "mlj017-all97-run-utf8.txt")
batch_path = os.path.join(eval_dir, "mlj017-smoke-v2-simple-pkg6gen-batch-input.json")
out_path = os.path.join(eval_dir, "mlj017-all97-results.md")

with open(raw_path, "r", encoding="utf-8") as f:
    content = f.read()

with open(batch_path, "r", encoding="utf-8") as f:
    batch = json.load(f)

q_map = {q["id"]: q["query"] for q in batch["questions"]}

# Split on separator lines (72 = signs)
SEP = "=" * 72

# Find all question blocks
results = []
# Split content into lines, filtering truly empty consecutive lines
lines = [l.rstrip() for l in content.split("\n")]

# State machine parse
i = 0
while i < len(lines):
    line = lines[i]
    # Look for separator
    if line == SEP:
        # Next non-blank line is label/id
        j = i + 1
        while j < len(lines) and lines[j].strip() == "":
            j += 1
        label_line = lines[j].strip() if j < len(lines) else ""
        id_match = re.match(r"^\[(\w+)\]", label_line)
        if not id_match:
            i += 1
            continue
        sq_id = id_match.group(1)
        query_start = j + 1
        # Skip blank lines and the next separator
        while query_start < len(lines) and (lines[query_start].strip() == "" or lines[query_start] == SEP):
            query_start += 1
        # query is the next non-blank non-sep line if it doesn't start with ---
        query = q_map.get(sq_id, label_line.replace(f"[{sq_id}]", "").strip(": ").strip())

        # Now scan for answer/sources/meta until next separator or end
        answer_lines = []
        source_lines = []
        elapsed_ms = 0
        domains = ""
        cache_hit = False
        timed_out = False

        section = None
        k = j + 1
        while k < len(lines):
            l = lines[k]
            if l == SEP and k + 1 < len(lines):
                # check if next non-blank is a new [sq...] label
                nk = k + 1
                while nk < len(lines) and lines[nk].strip() == "":
                    nk += 1
                if nk < len(lines) and re.match(r"^\[sq\d+\]", lines[nk].strip()):
                    break
            if "--- ANSWER ---" in l:
                section = "answer"
            elif "--- SOURCES ---" in l:
                section = "sources"
            elif "--- CITATIONS ---" in l:
                section = "citations"
            elif "--- META ---" in l:
                m = re.search(r"elapsed=(\d+)ms domains=(.+?) cacheHit=(\w+)", l)
                if m:
                    elapsed_ms = int(m.group(1))
                    domains = m.group(2).strip()
                    cache_hit = m.group(3) == "true"
                    timed_out = elapsed_ms >= 25000
                section = None
                k += 1
                break
            else:
                if section == "answer" and l.strip():
                    answer_lines.append(l.strip())
                elif section == "sources":
                    # Lines like "- filename.pdf"
                    sm = re.match(r"^- (.+\.(pdf|docx|xlsx|pptx))", l, re.IGNORECASE)
                    if sm:
                        source_lines.append(sm.group(1))
            k += 1

        results.append({
            "id": sq_id,
            "query": query,
            "answer": " ".join(answer_lines[:6]),
            "sources": source_lines,
            "elapsed_ms": elapsed_ms,
            "domains": domains,
            "cache_hit": cache_hit,
            "timed_out": timed_out,
        })
        i = k
    else:
        i += 1

print(f"Parsed {len(results)} results")

# Stats
total = len(results)
with_sources = sum(1 for r in results if r["sources"])
timed_out_list = [r["id"] for r in results if r["timed_out"]]
elapsed_vals = [r["elapsed_ms"] for r in results if r["elapsed_ms"] > 0]
avg_ms = sum(elapsed_vals) / len(elapsed_vals) if elapsed_vals else 0
max_ms = max(elapsed_vals) if elapsed_vals else 0
min_ms = min(elapsed_vals) if elapsed_vals else 0
total_sec = sum(elapsed_vals) // 1000

from datetime import date
today = date.today().isoformat()

lines_out = [
    "# MLJ-017 All-97 Eval Run Results",
    "",
    f"**Project:** MLJ-017 Package 6 - General `145b3dcf-272e-4c45-9e19-953f20f25bb9`",
    f"**Run date:** {today}",
    f"**Questions:** {total} (sq01–sq102, excl. sq32/sq50–sq53)",
    "**Hybrid retrieval:** ON · **Rerank:** OFF · **pgvector timeout:** 30s · **FTS timeout:** 25s",
    "**Input:** `eval/mlj017-smoke-v2-simple-pkg6gen-batch-input.json`",
    "",
    "---",
    "",
    "## Performance Summary",
    "",
    "| Metric | Value |",
    "|---|---|",
    f"| Total questions | {total} |",
    f"| Questions with sources returned | {with_sources} ({round(100*with_sources/total)}%) |",
    f"| Questions that hit DB timeout (≥25s) | {len(timed_out_list)} ({round(100*len(timed_out_list)/total)}%) |",
    f"| Min elapsed | {min_ms}ms |",
    f"| Max elapsed | {max_ms}ms |",
    f"| Avg elapsed | {round(avg_ms)}ms |",
    f"| Total wall time | {total_sec}s (~{round(total_sec/60)}min) |",
    "",
    f"**Timeout questions (fell back to keyword search):** {', '.join(timed_out_list)}",
    "",
    "---",
    "",
    "## Per-Question Results",
    "",
]

for r in results:
    elapsed_s = r["elapsed_ms"] / 1000
    if r["timed_out"]:
        timing = f"⏱️ TIMEOUT {elapsed_s:.1f}s (keyword fallback)"
    elif r["elapsed_ms"] < 1000:
        timing = f"✓ {r['elapsed_ms']}ms (exact-ID)"
    else:
        timing = f"{elapsed_s:.1f}s"

    src_display = ", ".join(f"`{s}`" for s in r["sources"][:3])
    if len(r["sources"]) > 3:
        src_display += f" +{len(r['sources'])-3} more"
    if not src_display:
        src_display = "_no sources_"

    answer_snippet = r["answer"][:400] + ("…" if len(r["answer"]) > 400 else "")

    lines_out += [
        f"### [{r['id']}] — {timing}",
        f"**Query:** {r['query']}",
        f"**Domains:** {r['domains'] or 'n/a'} | **Sources:** {src_display}",
        f"**Answer:** {answer_snippet}",
        "",
        "---",
        "",
    ]

output = "\n".join(lines_out)
with open(out_path, "w", encoding="utf-8") as f:
    f.write(output)

print(f"Report written: {out_path}")
print(f"Timed out ({len(timed_out_list)}): {', '.join(timed_out_list)}")

File: packages/backend/eval/generate-all97-report.ts
/**
 * Parse mlj017-all97-run.txt and produce a markdown results report.
 * Run: pnpm tsx ./eval/generate-all97-report.ts
 */
import fs from "node:fs";
import path from "node:path";

const evalDir = path.join(import.meta.dirname ?? __dirname);
const rawPath = path.join(evalDir, "mlj017-all97-run-utf8.txt");
const batchPath = path.join(evalDir, "mlj017-smoke-v2-simple-pkg6gen-batch-input.json");
const outPath = path.join(evalDir, "mlj017-all97-results.md");

interface Question {
  id: string;
  query: string;
  activeDocFileName?: string;
}

interface ParsedResult {
  id: string;
  query: string;
  answer: string;
  sources: string[];
  citations: string[];
  elapsedMs: number;
  domains: string;
  cacheHit: boolean;
  timedOut: boolean;
}

function parseOutput(raw: string, questions: Question[]): ParsedResult[] {
  // The UTF-16→UTF-8 conversion inserts blank lines between every real line; collapse them.
  const lines = raw.split(/\r?\n/).filter((l) => l.trim() !== "" || l === "");
  const results: ParsedResult[] = [];

  const questionMap = new Map<string, string>(questions.map((q) => [q.id, q.query]));

  let i = 0;
  while (i < lines.length) {
    // Look for separator lines (===)
    if (lines[i]?.match(/^={60,}$/)) {
      const labelLine = lines[i + 1]?.trim() ?? "";
      const queryLine = lines[i + 2]?.trim() ?? "";
      i += 3;

      // Extract id from label like "[sq01]" or "[sq12]: query text"
      const idMatch = labelLine.match(/^\[(\w+)\]/);
      if (!idMatch) continue;
      const id = idMatch[1];

      // If query is the next separator, we got label=id, query=labelLine content
      let query = queryLine;
      if (!query || query.match(/^={60,}$/)) {
        query = labelLine.replace(/^\[\w+\]:\s*/, "").replace(/^\[\w+\]/, "").trim();
      }
      if (!query) query = questionMap.get(id) ?? id;

      // Read answer section
      const answerLines: string[] = [];
      const sourceLines: string[] = [];
      const citationLines: string[] = [];
      let elapsedMs = 0;
      let domains = "";
      let cacheHit = false;
      let inAnswer = false, inSources = false, inCitations = false;

      while (i < lines.length) {
        const line = lines[i];
        if (line?.match(/^={60,}$/) && lines[i + 1]?.match(/^\[sq/)) break; // next question

        if (line?.startsWith("--- ANSWER ---")) { inAnswer = true; inSources = false; inCitations = false; i++; continue; }
        if (line?.startsWith("--- SOURCES ---")) { inAnswer = false; inSources = true; inCitations = false; i++; continue; }
        if (line?.startsWith("--- CITATIONS ---")) { inAnswer = false; inSources = false; inCitations = true; i++; continue; }
        if (line?.match(/^--- META --- elapsed=(\d+)ms domains=(.+?) cacheHit=(\w+)/)) {
          const m = line.match(/^--- META --- elapsed=(\d+)ms domains=(.+?) cacheHit=(\w+)/);
          if (m) { elapsedMs = parseInt(m[1]); domains = m[2].trim(); cacheHit = m[3] === "true"; }
          i++; break;
        }

        if (inAnswer && line !== undefined) answerLines.push(line);
        if (inSources && line?.match(/^- .+\.(pdf|docx|xlsx|pptx)/i)) sourceLines.push(line.replace(/^- /, "").trim());
        if (inCitations && line?.match(/^- .+\.(pdf|docx|xlsx|pptx)/i)) citationLines.push(line.replace(/^- /, "").trim());
        i++;
      }

      results.push({
        id,
        query,
        answer: answerLines.join("\n").trim(),
        sources: sourceLines,
        citations: citationLines,
        elapsedMs,
        domains,
        cacheHit,
        timedOut: elapsedMs >= 25000,
      });
    } else {
      i++;
    }
  }

  return results;
}

function buildReport(results: ParsedResult[], questions: Question[]): string {
  const now = new Date().toISOString().slice(0, 10);
  const total = results.length;
  const timedOut = results.filter((r) => r.timedOut).length;
  const cacheHits = results.filter((r) => r.cacheHit).length;
  const elapsedTimes = results.map((r) => r.elapsedMs).filter((t) => t > 0);
  const avgMs = elapsedTimes.reduce((a, b) => a + b, 0) / elapsedTimes.length;
  const maxMs = Math.max(...elapsedTimes);
  const minMs = Math.min(...elapsedTimes);
  const totalSec = Math.round(elapsedTimes.reduce((a, b) => a + b, 0) / 1000);

  const withSources = results.filter((r) => r.sources.length > 0).length;

  let md = `# MLJ-017 All-97 Eval Run Results

**Project:** MLJ-017 Package 6 - General \`145b3dcf-272e-4c45-9e19-953f20f25bb9\`
**Run date:** ${now}
**Questions:** ${total} (sq01–sq102, excl. sq32/sq50–sq53)
**Hybrid retrieval:** ON · **Rerank:** OFF
**Input:** \`eval/mlj017-smoke-v2-simple-pkg6gen-batch-input.json\`
**Raw output:** \`eval/mlj017-all97-run.txt\`

---

## Performance Summary

| Metric | Value |
|---|---|
| Total questions | ${total} |
| Questions with sources returned | ${withSources} (${Math.round(100*withSources/total)}%) |
| Questions that hit DB timeout (≥25s) | ${timedOut} (${Math.round(100*timedOut/total)}%) |
| Cache hits | ${cacheHits} |
| Min elapsed | ${minMs}ms |
| Max elapsed | ${maxMs}ms |
| Avg elapsed | ${Math.round(avgMs)}ms |
| Total wall time | ${totalSec}s (~${Math.round(totalSec/60)}m) |

---

## Per-Question Results

`;

  // Group by category from the batch questions
  // Build a map of id -> question
  const qMap = new Map<string, Question>(questions.map((q) => [q.id, q]));

  for (const r of results) {
    const elapsedLabel = r.timedOut ? ` ⏱️ TIMEOUT (${(r.elapsedMs/1000).toFixed(1)}s)` :
      r.elapsedMs < 1000 ? ` ✓ ${r.elapsedMs}ms (exact-ID)` :
      ` ${(r.elapsedMs/1000).toFixed(1)}s`;

    const sourcesLabel = r.sources.length > 0
      ? r.sources.slice(0, 3).map((s) => `\`${s}\``).join(", ") + (r.sources.length > 3 ? ` +${r.sources.length-3} more` : "")
      : "_no sources_";

    md += `### [${r.id}]${elapsedLabel}\n`;
    md += `**Query:** ${r.query}\n\n`;
    md += `**Domains:** ${r.domains || "n/a"}\n\n`;
    md += `**Sources:** ${sourcesLabel}\n\n`;

    // Truncate answer to first 400 chars
    const answerSnippet = r.answer.replace(/\n+/g, " ").slice(0, 400);
    if (answerSnippet) md += `**Answer snippet:** ${answerSnippet}${r.answer.length > 400 ? "…" : ""}\n\n`;

    md += `---\n\n`;
  }

  return md;
}

async function main() {
  const raw = fs.readFileSync(rawPath, "utf8");
  const batchJson = JSON.parse(fs.readFileSync(batchPath, "utf8")) as { questions: Question[] };
  const questions = batchJson.questions;

  console.log(`Parsing ${raw.split("\n").length} lines, ${questions.length} questions...`);
  const results = parseOutput(raw, questions);
  console.log(`Parsed ${results.length} results`);

  const timedOut = results.filter((r) => r.timedOut);
  console.log(`Timed out: ${timedOut.map((r) => r.id).join(", ")}`);

  const report = buildReport(results, questions);
  fs.writeFileSync(outPath, report, "utf8");
  console.log(`Report written to ${outPath}`);
}

main().catch(console.error);
